### PR TITLE
[release/3.1] Pass access token as query string when running SignalR in the browser

### PR DESCRIPTION
**Pull request by wtgodbe**
Resolves https://github.com/dotnet/aspnetcore/issues/18697

#### Description

When running in the browser, we can't use request headers. The SignalR client tries to do so anyways, which causes issues w/ Blazor Wasm. This PR detects if we're running in the browser, and sends the access token as a query string instead.

#### Customer Impact

Customers writing Blazor Wasm apps w/ SignalR will not see issues when adding Auth to their apps

#### Regression?

No

#### Risk

Behavior has been tested locally, but there could be obscure scenarios we've missed.

I used https://github.com/wtgodbe/scratch/tree/master/SigRBlazor/BlazorSignalRApp to validate that the token gets passed as a query string & can be pulled out and referenced by the server.

### CODE REVIEW

**Review by BrennanConroy — APPROVED** [review]

**Review by halter73 — APPROVED** [review]
Original PR #20115

File: src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/Utils.cs
Comment (davidfowl) [inline src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/Utils.cs:48]: Is this what we agreed on with the BCL team?

File: src/SignalR/clients/csharp/Http.Connections.Client/src/HttpConnection.cs
Comment (davidfowl) [inline src/SignalR/clients/csharp/Http.Connections.Client/src/HttpConnection.cs:380]: Why is that? Is it because the http client isn't fetch based?

File: src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/Utils.cs
Comment (wtgodbe) [reply src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/Utils.cs:48]: The guidance I got from @pranavkm and @anurse was that this is the pattern going forward

File: src/SignalR/clients/csharp/Http.Connections.Client/src/HttpConnection.cs
Comment (wtgodbe) [reply src/SignalR/clients/csharp/Http.Connections.Client/src/HttpConnection.cs:380]: From @BrennanConroy on the other PR:

> Was thinking about this over the weekend and realized we need to do something about SSE.

Right now it doesn't work because we use an HttpClient GET with "text/event-stream" Accept header. Which doesn't work in the browser, we use the EventSource API in the Typescript client to do a streaming response.

We might just have to disable SSE when running the .NET client in the browser.

File: src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/Utils.cs
Comment (anurse) [reply src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/Utils.cs:48]: This is what we have now. I believe discussions are ongoing and we can react later if necessary. Personally I’d prefer a string like “browser” since WASM is not browser-specific, but 🤷‍♀️

File: src/SignalR/clients/csharp/Http.Connections.Client/src/HttpConnection.cs
Comment (anurse) [reply src/SignalR/clients/csharp/Http.Connections.Client/src/HttpConnection.cs:380]: I believe @brennanconroy tried it and it didn’t work. We can always make adjustments and reenable. Plus, we are considering deprecating SSE in general.

File: src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/Utils.cs
Comment (davidfowl) [reply src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/Utils.cs:48]: The string needs to be browser. Can we circle back and make sure? Last I checked we were leaning in that direction. I worry about future breaking changes

Comment (anurse) [reply src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/Utils.cs:48]: @pranavkm @danroth27 @SteveSandersonMS What's the story here? Is there a final decision on the string to use for detecting the browser? We'd really like to get this in and right the first time before 3.1.4 closes on Tuesday.

Comment (pranavkm) [reply src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/Utils.cs:48]: It's changing to `BROWSER` See: https://github.com/mono/mono/pull/19482. Also an issue for this: https://github.com/dotnet/runtime/issues/33328#issuecomment-596072263

Comment (pranavkm) [reply src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/Utils.cs:48]: FYI @lewing

Comment (wtgodbe) [reply src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/Utils.cs:48]: Is there any idea when that'll happen? I guess we'll need to either wait, or change the string at the same time they change it

Comment (pranavkm) [reply src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/Utils.cs:48]: The change should land before 3.2-RC. Since this change aligns with the RTM release, using `BROWSER` might be the correct thing to do. Again @lewing in case you think otherwise.

Comment (anurse) [reply src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/Utils.cs:48]: We can change it now to `BROWSER`. The timing should be fine (3.1.4 is shipping no earlier than 3.2 GA, by the current schedule). It's not going to *fail* if the string is wrong, right? It just won't light up the fixed behavior until you're running on a runtime with the updated OS platform string.

Comment (wtgodbe) [reply src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/Utils.cs:48]: > It's not going to fail if the string is wrong, right? It just won't light up the fixed behavior until you're running on a runtime with the updated OS platform string.

That'd be my understanding, it won't fail, it just won't do anything until the corresponding change is made. I'll make the change

**Comment by wtgodbe** [comment]
Windows failures are https://github.com/dotnet/aspnetcore/issues/20687

**Comment by wtgodbe** [comment]
@anurse should we wait on getting the 3.1 CI fixed before merging? Failures are due to https://github.com/dotnet/aspnetcore/issues/20687, which I'm working on now

**Comment by anurse** [comment]
With https://github.com/dotnet/aspnetcore/pull/20736 merged, are we ready to merge this in (once the build passes)?

**Comment by wtgodbe** [comment]
> With #20736 merged, are we ready to merge this in (once the build passes)?

That was my plan, just want to get a green CI first

**Comment by danroth27** [comment]
🎉 🚀

**Pull request merged**
